File: services/analyzer/analyzers/layout.py
"""Layout analyzer: how many detected content blocks actually share an edge
(left/right/top/bottom) with another block, within tolerance. Low alignment
ratio is a concrete, measurable "things don't line up" signal."""

from __future__ import annotations

import cv2
import numpy as np

from models import Finding, Measured
from shared import TextRegion, detect_content_boxes, new_id

TOLERANCE_FRAC = 0.01  # edges within 1% of image width/height count as aligned
EXPECTED_ALIGNMENT_RATIO = (0.6, 1.0)


def _edges_aligned(a: float, b: float, tol: float) -> bool:
    return abs(a - b) <= tol


def analyze(bgr: np.ndarray, text_regions: list[TextRegion]) -> list[Finding]:
    h, w = bgr.shape[:2]
    gray = cv2.cvtColor(bgr, cv2.COLOR_BGR2GRAY)
    boxes = detect_content_boxes(gray)
    # Prefer OCR boxes when we have enough of them -- more reliable edges
    # than generic contour detection for text-heavy designs.
    if len(text_regions) >= 4:
        boxes = [tuple(int(v) for v in r.bbox) for r in text_regions]

    if len(boxes) < 3:
        return []

    tol_x = TOLERANCE_FRAC * w
    tol_y = TOLERANCE_FRAC * h

    aligned_count = 0
    for i, box in enumerate(boxes):
        x, y, bw, bh = box
        has_partner = False
        for j, other in enumerate(boxes):
            if i == j:
                continue
            ox, oy, ow, oh = other
            if (
                _edges_aligned(x, ox, tol_x)
                or _edges_aligned(x + bw, ox + ow, tol_x)
                or _edges_aligned(y, oy, tol_y)
            ):
                has_partner = True
                break
        if has_partner:
            aligned_count += 1

    ratio = aligned_count / len(boxes)
    lo, hi = EXPECTED_ALIGNMENT_RATIO
    if ratio >= lo:
        return []

    xs = [b[0] for b in boxes]
    ys = [b[1] for b in boxes]
    x2s = [b[0] + b[2] for b in boxes]
    y2s = [b[1] + b[3] for b in boxes]
    union_bbox = (min(xs), min(ys), max(x2s) - min(xs), max(y2s) - min(ys))

    severity = "critical" if ratio < lo * 0.5 else "major"
    return [
        Finding(
            id=new_id(),
            dimension="layout",
            severity=severity,
            bbox=union_bbox,
            measured=Measured(value=round(ratio, 2), expected=EXPECTED_ALIGNMENT_RATIO, unit=" alignment ratio"),
        )
    ]

File: src/lib/measure/analyzers/layout.ts
import { newFindingId } from "../id";
import type { TextLine } from "../text-lines";
import { selectBoxes } from "./_boxes";
import { SKIPPED, measured, type AnalyzerResult } from "./_result";

/**
 * Layout analyzer: how many detected content blocks actually share an edge
 * (left/right/top) with another block, within tolerance. Low alignment
 * ratio is a concrete, measurable "things don't line up" signal. Port of
 * `services/analyzer/analyzers/layout.py`.
 */
const TOLERANCE_FRAC = 0.01;
const EXPECTED_ALIGNMENT_RATIO: [number, number] = [0.6, 1.0];

function edgesAligned(a: number, b: number, tol: number): boolean {
  return Math.abs(a - b) <= tol;
}

export function analyzeLayout(
  gray: ArrayLike<number>,
  width: number,
  height: number,
  textLines: TextLine[],
): AnalyzerResult {
  const boxes = selectBoxes(gray, width, height, textLines);
  if (boxes.length < 3) return SKIPPED;

  const tolX = TOLERANCE_FRAC * width;
  const tolY = TOLERANCE_FRAC * height;

  let alignedCount = 0;
  for (let i = 0; i < boxes.length; i++) {
    const { x, y, w } = boxes[i];
    let hasPartner = false;
    for (let j = 0; j < boxes.length; j++) {
      if (i === j) continue;
      const o = boxes[j];
      if (
        edgesAligned(x, o.x, tolX) ||
        edgesAligned(x + w, o.x + o.w, tolX) ||
        edgesAligned(y, o.y, tolY)
      ) {
        hasPartner = true;
        break;
      }
    }
    if (hasPartner) alignedCount++;
  }

  const ratio = alignedCount / boxes.length;
  const [lo, hi] = EXPECTED_ALIGNMENT_RATIO;
  if (ratio >= lo) return measured();

  const xs = boxes.map((b) => b.x);
  const ys = boxes.map((b) => b.y);
  const x2s = boxes.map((b) => b.x + b.w);
  const y2s = boxes.map((b) => b.y + b.h);
  const minX = Math.min(...xs);
  const minY = Math.min(...ys);
  const unionBbox: [number, number, number, number] = [
    minX,
    minY,
    Math.max(...x2s) - minX,
    Math.max(...y2s) - minY,
  ];

  return measured([
    {
      id: newFindingId(),
      dimension: "layout",
      severity: ratio < lo * 0.5 ? "critical" : "major",
      bbox: unionBbox,
      measured: { value: Math.round(ratio * 100) / 100, expected: [lo, hi], unit: " alignment ratio" },
    },
  ]);
}
